pythonforandroid/recipes/python3/patches/fix-ctypes-util-find-library.patch

```diff
@@ -10,9 +10,15 @@ diff --git a/Lib/ctypes/util.py b/Lib/ctypes/util.py
 +if True:
 +    def find_library(name):
 +        # Check the user app libs and system libraries directory:
-+        app_root = os.path.normpath(os.path.abspath('../../'))
-+        lib_search_dirs = [os.path.join(app_root, 'lib'), "/system/lib"]
-+        for lib_dir in lib_search_dirs:
++        from jnius import autoclass
++        activity = autoclass('org.kivy.android.PythonActivity').mActivity
++        package_name = activity.getApplicationContext().getPackageName()
++        manager = activity.getApplicationContext().getPackageManager()
++        nativeLibDir = manager.getApplicationInfo(
++            package_name, 1024  # PackageManager.GET_SHARED_LIBRARY_FILES
++        ).nativeLibraryDir
```

Comment:
Thanks for the contribution!
If I understand correctly you mainly changed the "hardcoded" `os.path.join(app_root, 'lib')` path for something more dynamic using pyjnius right?
What's the ouput of `nativeLibDir` then? I think we could still ditch the necessity if making it dynamic and just adding more "hardcoded" path the the `lib_search_dirs`. I think it's common even for things like gcc to have a list of hardcoded search path including some that don't exist. I think it would make the code easier to follow to just have these path hardcoded

Reply:
@AndreMiras on android these can all be in different paths like `/mnt/` so hardcoding an absolute path is impossible (as you probably already know) and ARM64 lacks the "lib" symlink inside the files dir where the current working directory is so you would need to descend out and into something else that depends on the app name for a relative path (which is impossible to hardcode without a dynamic component, and seems super risky given how much internal android dirs can move around between versions). In conclusion I agree it's a good point, but not practical anymore. Also for what it's worth, `/system/lib` is still hardcoded

Here is some related info: https://stackoverflow.com/questions/41709129/lib-directory-missing-after-installing-64-bit-android-application

Reply:
Useful info, thanks for sharing!